src/deprecated.js

```diff
@@ -15,7 +15,13 @@ define( [
 
 // Support: Android <=4.0 only
 // Make sure we trim BOM and NBSP
-var rtrim = /^[\s\uFEFF\xA0]+|[\s\uFEFF\xA0]+$/g;
+// Require that the "whitespace run" starts from a non-whitespace
+// to avoid O(N^2) behavior when the engine would try matching "\s$" at each space position.
+// It is important that non-whitespace char is mandatory for regexp performance reasons,
+// however, it does not break correctness since whitespace-only string will be trimmed by ltrim above.
+var ltrim = /^[\s\uFEFF\xA0]+/,
+	rtrim = /([^\s\uFEFF\xA0])[\s\uFEFF\xA0]+$/,
```

Comment:
I'm wondering if we really need it if we use native `trim` by default. It'd be much shorter if the fallback logic stayed the same. I understand that'd mean the performance issues still exist in Android <=4.0 but that browser is hardly used at this moment so making everyone download extra bytes to resolve a performance issue that only exists there doesn't seem worth it to me.

Reply:
> I understand that'd mean the performance issues still exist in Android <=4.0 

Performance issues happen in ALL the browsers. For instance, 10 second trim was on my Apple M1 Max notebook.

---

The new jQuery 4 has no `jQuery.trim` at all, so I am not sure if it is worth discussing browser support here.

If you can somehow prove that all the browsers supported by jQuery3 have a good native trim implementation, then we can indeed drop the regular expressions.

However, I just wanted to play safe here: keep backward compatibility (jQuery is used by a LOT of apps!), and improve the performance for well-known cases (native trim, and the one without native trim).

Reply:
Just in case, IE9 does not seem to support trim natively: https://caniuse.com/mdn-javascript_builtins_string_trim, so the polyfill it might still be needed.

Reply:
> > I understand that'd mean the performance issues still exist in Android <=4.0
> 
> Performance issues happen in ALL the browsers. For instance, 10 second trim was on my Apple M1 Max notebook.

I don't think you're running Android 4.0 on your Apple M1 Max MacBook. :) Since native trim is preferred when it works, it would be applied everywhere except for Android 4.0. Therefore, the performance issue would only exist on that very old browser.

> The new jQuery 4 has no `jQuery.trim` at all, so I am not sure if it is worth discussing browser support here.

I don't understand this point. jQuery 3 & 4 have their own browser support matrices and they are of course relevant when discussing changes to land in either of those lines.

> Just in case, IE9 does not seem to support trim natively: [caniuse.com/mdn-javascript_builtins_string_trim](https://caniuse.com/mdn-javascript_builtins_string_trim), so the polyfill it might still be needed.

This info is incorrect. `trim` support in IE started with version 9, not 10.

> If you can somehow prove that all the browsers supported by jQuery3 have a good native trim implementation, then we can indeed drop the regular expressions.

But your PR always uses native trim as long as it doesn't have this Android 4.0 bug. If you found a browser with very slow native trim, your PR would still make it use that very slow trim. What does it have to do with the regular expressions used here?

Reply:
> But your PR always uses native trim as long as it doesn't have this Android 4.0 bug

I've no idea what was Android 4 bug, and the comments in code do not explain it.
So I assume something was wrong, and I just fix the regex so it is identical in the behavior.


>But your PR always uses native trim as long as it doesn't have this Android 4.0 bug. If you found a browser with very slow native trim, your PR would still make it use that very slow trim. What does it have to do with the regular expressions used here?

I assume native trim (if available) should always be faster.
However, I assume it is worth fixing the regex just in case there's a browser that does not support trim natively.

> Therefore, the performance issue would only exist on that very old browser.

I do not know if there are other browsers that fail to support trim natively, and if jQuery cares to support such browsers, then why make the support inherently slow?

Reply:
The jQuery browser support policy is defined at https://jquery.com/browser-support. We support those browsers and we don't really care about others. If they don't support `String#trim` then most likely they don't support a ton of other stuff that jQuery requires to work.